### PR TITLE
gitserver: disable sg maintenance

### PR DESCRIPTION
**Pull request by stefanhengl**
This PR stes GCAUTO as the new default and disables sg maintenance.

sg maintenance has probably caused more harm than good. While it does have a significant performance impact for very large repos, it is also those repos that suffered the most from repo corruption.

The job has been disabled on many customer instances and, as far as I know, on all managed instances.

## Test plan
- just a config change. The new default config already runs succesfuly on many customer instances.

### CODE REVIEW

**Comment by stefanhengl** [comment]
cc @DaedalusG

**Review by sashaostrikov — APPROVED** [review]
We will completely fix sgm one day and bring it back as a default again :)

**Review by keegancsmith — APPROVED** [review]

**Comment by mrnugget** [comment]
Since you've all gathered here, out of curiosity: what was the original reason for introducing sg maintenance? why is it better than GCAUTO?

**Comment by stefanhengl** [comment]
TLDR: it speeds up git-log operations for very large repos. EG, without it, displaying git history (last 5 commits of a file) becomes very slow. At the time we introduced this, git didn't expose the right flags/config on the porcelain level. I think it is worth re-checking this and maybe contribute to git rather than writing our own wrapper around plumbing commands.
 
Here is some backround
- [benchmarks](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/search-scratch/-/blob/2021/monorepo/sg_maintenance.org?subtree=true)
- [comments in the code](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/gitserver/server/sg_maintenance.sh?subtree=true)

**Comment by DaedalusG** [comment]
@mrnugget Additionally around the same time this was being developed we were trying to resolve an issue in which a customer was experiencing failures in indexing their monorepo, in which `zoekt` would fail to fetch from `gitserver`.
https://github.com/sourcegraph/customer/issues/586

**Comment by stefanhengl** [comment]
Ah, yes good data point. Limiting the memory with `-window-memory` definitely helped. 

I guess the decision is between which kinds of problems are more acceptable for customers. I believe repo corruption tops everything, especially for monorepo customers where re-cloning is expensive. By disabling sourcegraph maintenance we will get more complaints about MEM and slow UI. At least we know how to limit MEM, because -window-memory as exposed as config parameter.

**Comment by stefanhengl** [comment]
For anyone who wonders what that means with respect to their Sourcegraph installation:

This change will roll out with Sourcegraph 4.1.

Overall, this change will increase stability for all customers.

Customers with large monorepos have most likely already disabled sg maintenance, either because they ran into problems or because they were advised by their CE. 

For customers with large monorepos that have sg maintenance enabled, git operations based on git-log will most likely run slower now. An example is the history of the last commits displayed at the bottom of the tree view. The core search experience, however, is not affected. If they prefer to leave sg maintenance enabled (NOT RECOMMENDED), they have to explicitly set "SRC_ENABLE_SG_MAINTENANCE=true" and "SRC_ENABLE_SG_MAINTENANCE=false" for gitserver.

In some cases, gitserver might not be able to run garbage collection because of lack of memory. This might manifest itself by a failed fetch or clone. In this case we advise customers to set `git config pack.windowMemory 100m` either globally or for the repo in question on gitserver.

Customers with small to medium sized repositories won't notice a difference.
